### PR TITLE
WT-8958 Improvements to ARM and other stress tests on evergreen

### PR DESCRIPTION
**Pull request by ajmorton**
Re-delivery of #7707 after compilation errors were detected in Evergreen.

When `HAVE_DIAGNOSTIC=0` the code in `__wt_yield_no_barrier` becomes:
```
__wt_abort(NULL);
sched_yield();
```
which triggers an `unreachable-code` warning in clang. The same code does not trigger a warning in gcc, I think due to how each compiler handles the `noreturn` attribute set on `__wt_abort`.

The code in `__wt_yield_no_barrier` has been updated to avoid this issue and new tests have been added to `compile-gcc` and `compile-clang` that will catch this during PRs if it happens again.
The first commit is the code already reviewed in #7707, the subsequent commits are the corrections to address the issue.

### CODE REVIEW

**Comment by agorrod** [comment]
I'm watching this, but waiting for testing to turn green. Let me know if that's not the right choice.

**Comment by ajmorton** [comment]
> I'm watching this, but waiting for testing to turn green. Let me know if that's not the right choice.

Yep makes sense to me 👍 . 
I checked the failing test and it's a python test with no relation to the PR code changes. 
I expect this is related to a different ticket. I've restarted the test in question to make sure it's a transient failure and not consistent, should be finished in ~10 minutes.

File: src/os_posix/os_yield.c [hunk 1]
```diff
@@ -24,3 +24,29 @@ __wt_yield(void) WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
...
     sched_yield();
 }
+
+/*
+ * If HAVE_DIAGNOSTIC is 0 clang will raise an error when the noreturn attribute is missing, and if
+ * HAVE_DIAGNOSTIC is 1 clang will raise an error when the noreturn attribute is present. This
+ * function should never be called when HAVE_DIAGNOSTIC is 0 so we can ignore the missing-noreturn
+ * warning here.
+ */
+#pragma GCC diagnostic push
```
Comment (ajmorton) [inline src/os_posix/os_yield.c:34]: Suppressing `missing-noreturn` here isn't my favourite solution but it's the cleanest fix I've found so far.
Any ideas to improve on this are very welcome.

Comment (sulabhM) [reply src/os_posix/os_yield.c:34]: If we have two separate function definitions would it stop complaining:

#ifndef HAVE_DIAGNOSTIC
void
__wt_yield_no_barrier(void) WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
{
    __wt_abort(NULL);
}
#else
void
__wt_yield_no_barrier(void) WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
{
    sched_yield();
}
#endif

Comment (ajmorton) [reply src/os_posix/os_yield.c:34]: We hit some issues with `s_all` in that case. 
`python prototypes.py` adds both function signatures in `extern_posix.h` that only differ by the `noreturn` attribute, which causes a `-Wredundant-decls` compilation error

Comment (sulabhM) [reply src/os_posix/os_yield.c:34]: That is annoying. I will LGTM, but maybe we should fix s_all ?

**Review by sulabhM — APPROVED** [review]
Consider if it would make sense to fix s_all for the alternate solution ?

**Comment by ajmorton** [comment]
> Consider if it would make sense to fix s_all for the alternate solution ?

Unfortunately not without a rewrite of `prototypes.py`. The script uses regex to find function definitions and to handle the alternate solution we'd either need to add some very complex edge cases or to parse an Abstract Syntax Tree that's preprocessor aware, neither of which are easy.

We're technically able to wrap an `#ifdef` around the `noreturn` attribute, but the resulting code is really ugly:
```
void
__wt_yield_no_barrier(void)
#ifndef HAVE_DIAGNOSTIC
    WT_GCC_FUNC_ATTRIBUTE((noreturn))
#endif
    WT_GCC_FUNC_ATTRIBUTE((visibility("default")))
{
#ifndef HAVE_DIAGNOSTIC
    __wt_abort(NULL);
#else
    sched_yield();
#endif
}
```

**Comment by ajmorton** [comment]
One last note for peoples interest; The reason gcc doesn't report an `unreachable-code` warning is it has been [removed from gcc](https://gcc.gnu.org/legacy-ml/gcc-help/2011-05/msg00360.html). 
@agorrod I haven't found a better implementation for the comment above and I'm going to call it here. Let me know if you'd like to review or I can merge as is.

**Review by agorrod — APPROVED** [review]
lgtm

**Pull request merged**